cpp/src/io/parquet/rle_stream.cuh

```diff
@@ -0,0 +1,359 @@
+template <typename level_t>
+struct rle_stream {
+  int level_bits;
+  uint8_t const* start;
+  uint8_t const* cur;
+  uint8_t const* end;
+
+  int max_output_values;
+  int total_values;
+  int cur_values;
+
+  level_t* output;
+
+  rle_run<level_t>* runs;
+  int run_index;
+  int run_count;
+  int output_pos;
+  bool spill;
+
+  int next_batch_run_start;
+  int next_batch_run_count;
+
+  __device__ rle_stream(rle_run<level_t>* _runs) : runs(_runs) {}
+
+  __device__ void init(int _level_bits,
+                       uint8_t const* _start,
+                       uint8_t const* _end,
+                       int _max_output_values,
+                       level_t* _output,
+                       int _total_values)
+  {
+    level_bits = _level_bits;
+    start      = _start;
+    cur        = _start;
+    end        = _end;
+
+    max_output_values = _max_output_values;
+    output            = _output;
+
+    run_index            = 0;
+    run_count            = 0;
+    output_pos           = 0;
+    spill                = false;
+    next_batch_run_start = 0;
+    next_batch_run_count = 0;
+
+    total_values = _total_values;
+    cur_values   = 0;
+  }
+
+  __device__ inline thrust::pair<int, int> get_run_batch()
+  {
+    return {next_batch_run_start, next_batch_run_count};
+  }
+
+  // fill in up to num_rle_stream_decode_warps runs or until we reach the max_count limit.
+  // this function is the critical hotspot.  please be very careful altering it.
+  __device__ inline void fill_run_batch(int max_count)
+  {
+    // if we spilled over, we've already got a run at the beginning
+    next_batch_run_start = spill ? run_index - 1 : run_index;
+    spill                = false;
+
+    // generate runs until we either run out of warps to decode them with, or
+    // we cross the output limit.
+    while (run_count < num_rle_stream_decode_warps && output_pos < max_count && cur < end) {
+      auto& run = runs[rolling_run_index(run_index)];
+
+      // Encoding::RLE
+
+      // bytes for the varint header
+      uint8_t const* _cur = cur;
+      int const level_run = get_vlq32(_cur, end);
+      int run_bytes       = _cur - cur;
+
+      // literal run
+      if (level_run & 1) {
+        int const run_size  = (level_run >> 1) * 8;
+        run.size            = run_size;
+        int const run_size8 = (run_size + 7) >> 3;
+        run_bytes += run_size8 * level_bits;
+      }
+      // repeated value run
+      else {
+        run.size = (level_run >> 1);
+        run_bytes++;
+        // can this ever be > 16?  it effectively encodes nesting depth so that would require
+        // a nesting depth > 64k.
+        if (level_bits > 8) { run_bytes++; }
+      }
+      run.output_pos = output_pos;
+      run.start      = _cur;
+      run.level_run  = level_run;
+      run.remaining  = run.size;
+      cur += run_bytes;
+
+      output_pos += run.size;
+      run_count++;
+      run_index++;
+    }
+
+    next_batch_run_count = run_count;
```

Comment:
AFAICT, next_batch_run_start and next_batch_run_count are used to signal the remaining kernels that more runs are ready for decoding. I'd find this code easier to follow if this was emphasized in some way.